runtime/src/main/java/org/corfudb/runtime/LogReplicationUtils.java

```diff
@@ -1,14 +1,269 @@
+                        // to set the baseline at this timestamp.
+                        clientListener.mergeTableOnInitialSubscription(multiTableReadTimestamp);
+
+                        // Return this timestamp.  The client's listener must be subscribed at this timestamp.
+                        return multiTableReadTimestamp;
+                    } catch (Exception e) {
+                        log.error("Error while attempting to get a valid subscription timestamp.  Retrying.", e);
+                        throw new RetryNeededException();
+                    }
+                }).setOptions(x -> {
+                    x.setBackoffDuration(Duration.ofMillis(backoff_retry_duration_ms));
+                    x.setMaxRetryThreshold(Duration.ofMinutes(max_duration_threshold_mins));
+                }).run();
+            } catch (InterruptedException e) {
+                log.error("Failed to get a valid subscription timestamp.", e);
+                clientListener.onError(e);
+            }
+            return null;
+        }
+
+        /**
+         * Gets the latest token from the sequencer and builds a timestamp from it
+         * @return Timestamp
+         */
+        private CorfuStoreMetadata.Timestamp getTimestamp() {
+            Token token = corfuStore.getRuntime().getSequencerView().query().getToken();
+            return CorfuStoreMetadata.Timestamp.newBuilder()
+                .setEpoch(token.getEpoch())
+                .setSequence(token.getSequence())
+                .build();
+        }
+
+        /**
+         * Checks if snapshot sync is ongoing at the given timestamp.
+         * @param timestamp
+         * @return
+         */
+        private boolean checkSnapshotSyncOngoing(CorfuStoreMetadata.Timestamp timestamp) {
+            try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE, IsolationLevel.snapshot(timestamp))) {
+                List<CorfuStoreEntry<ReplicationStatusKey, ReplicationStatusVal, Message>> entries =
+                    txn.executeQuery(REPLICATION_STATUS_TABLE, p -> true);
+                // In LR V1, it is a valid assumption that the size of replication status table will be 1 as there is
+                // only 1 remote cluster.  This implementation will change in LR V2
+                Preconditions.checkState(entries.size() == 1);
+
+                CorfuStoreEntry<ReplicationStatusKey, ReplicationStatusVal, Message> entry = entries.get(0);
+                txn.commit();
+                return entry.getPayload().getDataConsistent() ? false : true;
+            } catch (TransactionAbortedException e) {
+                // Since this is a read-only transaction, the abort can only be if this version of the table is not
+                // available in the JVM's cache.  The underlying cause for this error is TrimmedException
+                Preconditions.checkState(e.getCause() instanceof TrimmedException);
```

Comment:
Can we have the msg to print if the state is not true? Can be helpful to see what exception other than trimmedException occurred.